File: perl/shuffle-line/challenge.pl
#!/usr/bin/env perl
#
use strict;
use warnings;

sub shuffleLine {
  my ($names, $n) = @_;
  return @$names if $n <= 0;
  my @result = ();
  my @endOfLine = ();
  for (my $i = 0; $i < @$names; $i++) {
    if (($i + 1) % $n == 0) {
      push @endOfLine, $names->[$i];
    } else {
      push @result, $names->[$i];
    }
  }
  push @result, @endOfLine;
  return @result;
}

if (!caller()) {
  if (@ARGV != 2) {
    print STDERR "Usage: perl challenge.pl \"<csv-names>\" <integer-n>\n";
    exit 1;
  }

  my ($names_string, $n) = @ARGV;
  my @names = split(/,/, $names_string);
  @names = map { s/^\s+|\s+$//g; $_ } @names;
  print "# given names=[" . join(',', @names) . "] and n=$n, result is:\n";

  my @result = shuffleLine(\@names, $n);
  print "[" . join(',', @result) . "]\n";
}

1;

File: ruby/longest-sorted/challenge.rb
#!/usr/bin/env ruby

class Challenge
  attr_accessor :sentence

  def initialize(sentence)
    self.sentence = sentence
  end

  def words
    sentence.split.map { |word| word.downcase.gsub(/[^a-z]/, '') }
  end

  def alphabetically_sorted_words
    words.select { |word| word.length > 1 && word == word.chars.sort.join }
  end

  def longestSorted
    alphabetically_sorted_words.max_by { |word| word.length  } || ""
  end
end

if __FILE__ == $PROGRAM_NAME
  (puts "Usage: ruby #{$0} <sentence>"; exit) unless ARGV.length == 1
  sentence = ARGV[0]
  calculator = Challenge.new(sentence)
  puts "Sentence: #{calculator.sentence.inspect}"
  puts "Result: #{calculator.longestSorted.inspect}"
end
